sdks/python/apache_beam/runners/dataflow/dataflow_metrics.py

```diff
@@ -21,13 +21,101 @@
 
 
 # TODO(pabloem)(JIRA-1381) Implement this once metrics are queriable from
 # dataflow service
 class DataflowMetrics(MetricResults):
+  """Implementation of MetricResults class for the Dataflow runner."""
+
+  def __init__(self, dataflow_client=None, job_result=None):
+    """Initialize the dataflow metrics object.
+
+    Args:
+      dataflow_client: apiclient.DataflowApplicationClient to interact with the
+        dataflow serive.
+      job_result: DataflowPipelineResult with the state and id information of
+        the job
+    """
+    super(DataflowMetrics, self).__init__()
+    self._dataflow_client = dataflow_client
+    self.job_result = job_result
+    self._queried_after_termination = False
+
+  def _populate_metric_results(self, response):
+    """Take a list of metrics, and convert it to a list of MetricResult."""
+    user_metrics = [metric
+                    for metric in response.metrics
+                    if metric.name.origin == 'user']
+
+    # Get the tentative/committed versions of every metric together.
+    metrics_by_name = defaultdict(lambda: {})
+    for metric in user_metrics:
+      tentative = [prop
+                   for prop in metric.name.context.additionalProperties
+                   if prop.key == 'tentative' and prop.value == 'true']
+      key = 'tentative' if tentative else 'committed'
+      metrics_by_name[metric.name.name][key] = metric
+
+    # Now we create the MetricResult elements.
+    result = []
+    for name, metric in metrics_by_name.iteritems():
+      if (name.endswith('(DIST)') or
+          name.endswith('[MIN]') or
+          name.endswith('[MAX]') or
+          name.endswith('[MEAN]') or
+          name.endswith('[COUNT]')):
+        # Distributions are not yet fully supported in this runner
+        continue
+      [step, namespace, name] = name.split('/')
+      key = MetricKey(step, MetricName(namespace, name))
+      attempted = metric['tentative'].scalar.integer_value
+      committed = metric['committed'].scalar.integer_value
+      result.append(MetricResult(key, attempted=attempted, committed=committed))
+
+    return result
+
+  def _should_query_dataflow(self):
+    """Determine whether to query metrics from dataflow or use cached metrics.
+
+    If the last query was made after the job terminated, then it will use
+    the cached result. OTHERWISE, it should query the dataflow service.
+
+    Returns:
+      True if the dataflow service should be queried for metrics. False if the
+      cached query is good enough.
+    """
+    if self._queried_after_termination:
+      return False
+    else:
+      return True
+
+  def _get_metrics_from_dataflow(self):
+    """Return cached metrics or query the dataflow service."""
+    try:
+      job_id = self.job_result.job_id()
+    except AttributeError:
+      job_id = None
+    if not job_id:
+      raise ValueError('Can not query metrics. Job id is unknown.')
+
+    if self._should_query_dataflow():
```

Comment:
Can we make this logic simpler like this and remove _should_query_dataflow()?:

```python
if self._cached_metrics:
  return self._cached_metrics

job_metrics = self._dataflow_client.get_job_metrics(job_id)
# If the job has terminated, metrics will no longer change and we can cache the result.
if self.job_result.is_in_terminal_state():
  self._cached_metrics = job_metrics
```

In this case you also have to initialize `self._cached_metrics = None` in the constructor.